src/tiledbsoma_ml/_io_batch_iterable.py

```diff
@@ -0,0 +1,141 @@
+# Copyright (c) 2021-2024 The Chan Zuckerberg Initiative Foundation
+# Copyright (c) 2021-2024 TileDB, Inc.
+#
+# Licensed under the MIT License.
+
+import gc
+import logging
+import time
+from typing import Iterable, Iterator, Optional, Sequence, Tuple
+
+import attrs
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+from tiledbsoma import DataFrame, IntIndexer, SparseNDArray
+
+from tiledbsoma_ml._common import NDArrayJoinId
+from tiledbsoma_ml._csr import CSR_IO_Buffer
+from tiledbsoma_ml._eager_iter import EagerIterator
+from tiledbsoma_ml._query_ids import Chunks
+from tiledbsoma_ml._utils import batched
+
+logger = logging.getLogger("tiledbsoma_ml._io_batch_iterable")
+IOBatch = Tuple[CSR_IO_Buffer, pd.DataFrame]
+"""Tuple type emitted by |IOBatchIterable|, containing ``X`` rows (as a |CSR_IO_Buffer|) and ``obs`` rows
+(|pd.DataFrame|)."""
+
+
+@attrs.define(frozen=True)
+class IOBatchIterable(Iterable[IOBatch]):
+    """Given a list of ``obs_joinid`` |Chunks|, re-chunk them into (optionally shuffled) |IOBatch|'s".
+
+    An |IOBatch| is a tuple consisting of a batch of rows from the ``X`` |SparseNDArray|, as well as the corresponding
+    rows from the ``obs`` |DataFrame|. The ``X`` rows are returned in an optimized |CSR_IO_Buffer|.
+    """
+
+    chunks: Chunks
+    io_batch_size: int
+    obs: DataFrame
+    var_joinids: NDArrayJoinId
+    X: SparseNDArray
+    obs_column_names: Sequence[str] = ("soma_joinid",)
+    seed: Optional[int] = None
+    shuffle: bool = True
+    use_eager_fetch: bool = True
+
+    @property
+    def io_batch_ids(self) -> Iterable[Tuple[int, ...]]:
+        """Re-chunk ``obs_joinids`` according to the desired ``io_batch_size``."""
+        return batched(
+            (joinid for chunk in self.chunks for joinid in chunk),
+            self.io_batch_size,
+        )
+
+    def __iter__(self) -> Iterator[IOBatch]:
+        """Emit |IOBatch|'s."""
+        # Because obs/var IDs have been partitioned/split/shuffled upstream of this class, this RNG does not need to be
+        # identical across sub-processes, but seeding is supported anyway, for testing/reproducibility.
+        shuffle_rng = np.random.default_rng(self.seed)
+        X = self.X
+        context = X.context
+        obs_column_names = (
+            list(self.obs_column_names)
+            if "soma_joinid" in self.obs_column_names
+            else ["soma_joinid", *self.obs_column_names]
+        )
+        # NOTE: `.astype("int64")` works around the `np.int64` singleton failing reference-equality after cross-process
+        # serde.
```

Comment:
serde?  Did you mean [serdes](https://en.wikipedia.org/wiki/SerDes)?

Reply:
Huh, I've always seen it as "serde," but [SerDes](https://en.wikipedia.org/wiki/SerDes) seems clearest / most standard, so I went with that, thx.